### DISCUSSION TITLE
Auto complete

### DISCUSSION THREAD

**Discussion by MpouhaerB** [TomSchimansky/CustomTkinter#1641] [Q&A]
Hello,

Is there a way to customise a CTK text widget with auto-complete ?

For example with a list of values contained in a dataframe column ?

Thanks for your help,

**Comment by Akascape** [comment]
@MpouhaerB 
Try this: https://github.com/Akascape/CTkScrollableDropdown

**Reply by MpouhaerB** [reply]
Thanks ! I will explore this solution :)

**Reply by MpouhaerB** [reply]
@Akascape 

I'm on a Mac and when I try to implement the code for live search on VS Code, I get this error : TclError: expected boolean value but got ""

Here is my code with an entry : 

entry = customtkinter.CTkEntry(master=root,
                               placeholder_text="Entrez le titre d'un film",
                               width=200,
                               height=32,
                               border_width=1,
                               corner_radius=10)
entry.place(relx=0.45 + 0.01, rely=0.3, anchor='w')


dropdown = CTkScrollableDropdown(attach=entry,
                                 width=200,
                                 height=32,
                                 values=rec_engine_df['primaryTitle'].tolist(),
                                 command=lambda e: entry.insert(1, e),
                                 autocomplete=True) 

Can you help with this ? Thanks

**Reply by Akascape** [reply]
@MpouhaerB line number?

**Reply by MpouhaerB** [reply]
Sorry, here's the full error message : 

---------------------------------------------------------------------------
TclError                                  Traceback (most recent call last)
[~/Documents/Python/Interface](https://vscode-interactive+.vscode-resource.vscode-cdn.net//~/Documents/Python/Interface) projet 2[/Interface.py](https://vscode-interactive+.vscode-resource.vscode-cdn.net/Interface.py) in 
    [116](file:///Users/mickaelpouhaer/Documents/Python/Interface%20projet%202/Interface.py?line=115) 
    [117](file:///Users/mickaelpouhaer/Documents/Python/Interface%20projet%202/Interface.py?line=116) 
--> [118](file:///Users/mickaelpouhaer/Documents/Python/Interface%20projet%202/Interface.py?line=117) dropdown = CTkScrollableDropdown(attach=entry,
    [119](file:///Users/mickaelpouhaer/Documents/Python/Interface%20projet%202/Interface.py?line=118)                                  width=200,
    [120](file:///Users/mickaelpouhaer/Documents/Python/Interface%20projet%202/Interface.py?line=119)                                  height=32,

[~/Documents/Python/Interface](https://vscode-interactive+.vscode-resource.vscode-cdn.net//~/Documents/Python/Interface) projet 2[/ctk_scrollable_dropdown.py](https://vscode-interactive+.vscode-resource.vscode-cdn.net/ctk_scrollable_dropdown.py) in __init__(self, attach, x, y, button_color, height, width, fg_color, button_height, justify, scrollbar_button_color, scrollbar, scrollbar_button_hover_color, frame_border_width, values, command, image_values, alpha, frame_corner_radius, double_click, resize, frame_border_color, text_color, autocomplete, **button_kwargs)
     33             self.bind('', lambda e: self.withdraw() if not self.disable else None)
     34         elif sys.platform.startswith("darwin"):
---> 35             self.overrideredirect(True)
     36             self.transparent_color = 'systemTransparent'
     37             self.attributes("-transparent", True)

[~/opt/anaconda3/lib/python3.9/tkinter/__init__.py](https://vscode-interactive+.vscode-resource.vscode-cdn.net//~/opt/anaconda3/lib/python3.9/tkinter/__init__.py) in wm_overrideredirect(self, boolean)
   2174         if BOOLEAN is given with 1. Return the current value if None
   2175         is given."""
-> 2176         return self._getboolean(self.tk.call(
   2177             'wm', 'overrideredirect', self._w, boolean))
   2178 

[~/opt/anaconda3/lib/python3.9/tkinter/__init__.py](https://vscode-interactive+.vscode-resource.vscode-cdn.net//~/opt/anaconda3/lib/python3.9/tkinter/__init__.py) in _getboolean(self, string)
   1446         """Internal function."""
   1447         if string:
-> 1448             return self.tk.getboolean(string)
   1449 
   1450     def _displayof(self, displayof):

**Comment by Akascape** [comment]
@MpouhaerB In my mac, it is working fine. Try removing line 23: `self.attributes("-alpha", 0)`, Also try basic example and let me know if it's working.

**Reply by MpouhaerB** [reply]
I tried, I don't have the error anymore but the Python interface is crashing, unfortunately. However, your examples work fine.

**Reply by Akascape** [reply]
@MpouhaerB I think it is something to do with import method, use it like this `from CTkScrollableFrame import CTkScrollableFrame` and move it to top.

**Reply by Akascape** [reply]
@MpouhaerB Secondly this is actually a bug in tkinter itself (for mac). I have researched it, the `wm_overrideredirect()` method has some internal bugs if we use other attributes. Maybe a version difference is creating this.

**Reply by Akascape** [reply]
I was also getting this error with a different widget I made: https://github.com/Akascape/CTkPopupKeyboard, then I removed that attribute line and it worked fine.

**Reply by MpouhaerB** [reply]
Sorry, I'm a noob. What attribute line should I remove ? Also `from CTkScrollableFrame import CTkScrollableFrame` is not working for me. Here is what's working on top of my code : 

```
import customtkinter
from ctk_scrollable_dropdown import CTkScrollableDropdown
from customtkinter import CTkScrollableFrame
```

**Reply by Akascape** [reply]
@MpouhaerB Its not your fault, tkinter in mac is buggy itself. 
Make sure you have the full CTkScrollableDropdown folder,

```python
import customtkinter
from CTkScrollableDropdown import CTkScrollableDropdown
from customtkinter import CTkScrollableFrame
```
As I said, this error is related to some import methods.

**Reply by MpouhaerB** [reply]
The error message still shows. Never mind, thanks for your help ! I'll try to find a way :) I searched and some people had the same issue with PySimpleGUI : https://github.com/PySimpleGUI/PySimpleGUI/issues/3048, I tried to remove the line with the call to overrideredirect for Mac, I don't see the error message but Python is crashing.

**Reply by Akascape** [reply]
@MpouhaerB Hey, I just found the solution to the crash issue. Please update to new version: https://github.com/Akascape/CTkScrollableDropdown

**Reply by MpouhaerB** [reply]
Hey, I updated the version but Python is crashing without any error message.

**Reply by MpouhaerB** [reply]
Ha, and I also tried your example script and the widgets are no longer clickable.

**Reply by Akascape** [reply]
@MpouhaerB it's so strange, in my mac it is working fine.
Can you also try this widget: https://github.com/Akascape/CTkPopupKeyboard
If this one crashes, then I am sure it is something to do with python version.
My python version is 3.10

**Comment by MpouhaerB** [comment]
```
def show_popup():
    # Disable/Enable popup
    if switch.get()==1:
        keyboard.disable = False
        numpad.disable = False
    else:
        keyboard.disable = True
        numpad.disable = True

root = customtkinter.CTk()
root.title("Cine Creuse : vos recommandations de films")
root.geometry("1280x720")


entry = customtkinter.CTkEntry(master=root,
                               placeholder_text="Entrez le titre d'un film",
                               width=200,
                               height=32,
                               border_width=1,
                               corner_radius=10)
entry.place(relx=0.45 + 0.01, rely=0.2, anchor='w')


# dropdown = CTkScrollableDropdown(attach=entry,
#                                  width=200,
#                                  height=32,
#                                  values=rec_engine_df['primaryTitle'].tolist(),
#                                  command=lambda e: entry.insert(1, e),
#                                  autocomplete=True)

numpad = PopupNumpad(entry)



switch = customtkinter.CTkSwitch(root, text="On-Screen Keyboard", command=show_popup)
switch.pack(pady=10)
switch.toggle()

button = customtkinter.CTkButton(master=root,
                                 width=120,
                                 height=32,
                                 border_width=0,
                                 corner_radius=8,
                                 text="Rechercher",
                                 fg_color='red',
                                 hover_color='grey',
                                 command=search_movies)
button.place(relx=0.45, rely=0.3, anchor='e')
```
I have the switch but the keyboard is not showing up.

**Reply by Akascape** [reply]
Can you update python to 3.10 or newer?

**Reply by MpouhaerB** [reply]
I updated to 3.10, I still have the same issue with Python crashing and on the other hand your examples no longer being clickable. I have a conda environment maybe it's an explanation...

**Reply by Akascape** [reply]
@MpouhaerB 
> I have a conda environment maybe it's an explanation...

This can be the reason, I was actually testing with the python launcher and it works fine there.

**Reply by Akascape** [reply]
@MpouhaerB For windows, overrideredirect works properly.
For linux, using `.attribites("-type", "splash")` works fine instead of overrideredirect.

But for mac, that method disables all event bindings and attributes which is a big issue. Idk if there is any other method to make that window for mac.

**Reply by MpouhaerB** [reply]
Thanks for your help, I've found a work around vis CTkCombobox, it's not auto complete live search but I can generate a list of values in the combobox that are similar to the value inserted in my entry.

**Reply by Akascape** [reply]
@MpouhaerB How did you fix the issue? Just wanted to know what changes you made...

**Reply by MpouhaerB** [reply]
I didn't fix the issue, I've found a downgraded way to meet my need. From an entry, I generate a list of similar strings from my dataframe column and send it to the combobox.
